### ISSUE TITLE
Strapi v4, postgres, Error "Knex: Timeout acquiring a connection. The pool is probably full. Are you missing a .transacting(trx) call?"

### ISSUE THREAD

**Issue by smaranjit** [strapi/strapi#11860]
### Describe the bug

I am getting error message on Postgres database (version 14) but working perfectly with MySQL database
```
[2021-12-09 19:26:58.233] debug: ⛔️ Server wasn't able to start properly.
[2021-12-09 19:26:58.235] error: Knex: Timeout acquiring a connection. The pool is probably full. Are you missing a .transacting(trx) call?
KnexTimeoutError: Knex: Timeout acquiring a connection. The pool is probably full. Are you missing a .transacting(trx) call?
    at Client_PG.acquireConnection (/home/xxxx/xxx/node_modules/knex/lib/client.js:295:26)
    at runNextTicks (internal/process/task_queues.js:62:5)
    at listOnTimeout (internal/timers.js:523:9)
    at processTimers (internal/timers.js:497:7)
error Command failed with exit code 1.
```
when I have 34+ collections, 200+ relation types, 250+ tables in the database and when adding, removing or modifying field or relation

### Steps to reproduce the behavior

1. 34+ collections, 200+ relation type, 250+ tables in database 
2. modify any schema
3. See error

### Expected behavior

the server should restart without error

### System

- Node.js version: Node LTS version v14
- NPM version: npm v6.14.15
- Strapi version: v4
- Database: postgres
- Operating system: ubuntu 21.10

### Additional context
working fine for less no of collections or relations.

**Comment by scoorpionx** [comment]
Did you configure the connection pool?
https://docs.strapi.io/developer-docs/latest/setup-deployment-guides/configurations/required/databases.html#database-pooling-options

**Comment by smaranjit** [comment]
> Did you configure the connection pool? https://docs.strapi.io/developer-docs/latest/setup-deployment-guides/configurations/required/databases.html#database-pooling-options

yes, I have tried with different pool settings but nothing worked. but for MySQL, it's working without any pool settings.

**Comment by ihmpavel** [comment]
We had same issue on Postgres. It was because duplicated name/id. The only way to restore everything was resetting production database and importing data from scratch. 

It was because incorrected database model, which was exhausting our database (but never ended).

**Comment by derrickmehaffy** [comment]
Most likely not a bug but an environment issue. Recommend continuing discussion on our discord for forum.

**Comment by smaranjit** [comment]
> We had same issue on Postgres. It was because duplicated name/id. The only way to restore everything was resetting production database and importing data from scratch.
> 
> It was because incorrected database model, which was exhausting our database (but never ended). Only on production. Same settings - on localhost was working fine.

for me, it is happening in my localhost

**Comment by smaranjit** [comment]
> Most likely not a bug but an environment issue. Recommend continuing discussion on our discord for forum.

I think this is a bug, just think how frustrating the situation will be when we have to delete and recreate and refresh the whole database for any minor model changes when customizing the backend. There should be some warning or some properly described error, like which model or which id is conflicting. Instead of that, it just shows that generic error, which is very hard to debug or find out actually what is causing the problem,

**Comment by ihmpavel** [comment]
It was very frustrating. I tried everything. Only one version from backup was working, but newer version with slightly changed models (eg.: just new name) destroyed everything again. 

The issue was created before and happened only after some changes... I really do not wish anyone to experience same issues...

**Comment by smaranjit** [comment]
> It was very frustrating. I tried everything. Only one version from backup was working, but newer version with slightly changed models (eg.: just new name) destroyed everything again.
> 
> The issue was created before and happened only after some changes... I really do not wish anyone to experience same issues...

do we have any solution right now, other than recreating the database?

**Comment by reijerteunis** [comment]
I have the same issue, so frustrating. What to do?

**Comment by ihmpavel** [comment]
Our solution:

Pooling (to limit number of sql queries) did not work on server (for us). Our solution is to dump the database from production. Import it on local machine. Run `strapi develop`, export the dump from local machine. Upload the database dump (before the application itself) on production. After that we deploy the application to production and start it. It is working, but really frustrating. our database has a lot of tables (not data).

**Comment by Pablo-Aldana** [comment]
I'm having same issue when trying to deploy for the first time into PROD with empty DB. is this issue related just to Postgres?

**Comment by smaranjit** [comment]
> I'm having same issue when trying to deploy for the first time into PROD with empty DB. is this issue related just to Postgres?

for me, it seems like this issue is occurring for Postgres only, cause with the same backend when doing the same operation with the MySQL database then there is no error. probably there is some problem with the way `strapi4` talks with the `pg` npm library.

**Comment by derrickmehaffy** [comment]
> > I'm having same issue when trying to deploy for the first time into PROD with empty DB. is this issue related just to Postgres?
> 
> for me, it seems like this issue is occurring for Postgres only, cause with the same backend when doing the same operation with the MySQL database then there is no error. probably there is some problem with the way `strapi4` communicates with the `pg` npm library. 

Every pull request we do has end to end testing on every database we support including pg and MySQL.

There is no known issue on pg in our tests not my own personal testing on v4. The issue is almost certainly environmental.

**Comment by ari-motors-team** [comment]
I think, I am facing a similar issue.
#12086

**Comment by smaranjit** [comment]
> > > I'm having same issue when trying to deploy for the first time into PROD with empty DB. is this issue related just to Postgres?
> > 
> > 
> > for me, it seems like this issue is occurring for Postgres only, cause with the same backend when doing the same operation with the MySQL database then there is no error. probably there is some problem with the way `strapi4` communicates with the `pg` npm library.
> 
> Every pull request we do has end to end testing on every database we support including pg and MySQL.
> 
> There is no known issue on pg in our tests not my own personal testing on v4. The issue is almost certainly environmental.

But for almost the same type of schema structure, the backend is working fine with the same Postgres environment in Strapi-v3.6.8, but the problem is occurring for Starpi - v4, so it is very unlikely that it is an environmental issue. and also want to mention that this is not happening when there is a very small no of tables or relations, but it's happening when no of tables and relations increases, for me 250+ tables in Postgres. also on another note, the issue may come from anywhere but if there is no clear error message display, instead of that generic Knex error message, then it's very hard to find the real cause of the error.

**Comment by derrickmehaffy** [comment]
@ari-motors-team and @smaranjit then we need detailed reproduction steps.

A sample application or something that can be tested by us (ideally an application without much customization just native strapi stuff like models and what not, eg no uses of strapi.db in lifecycles or custom controllers).

My guess is the default configuration that some cloud hosts are applying is a conflict but without a means of reproduction there isn't much we can offer here.

**Comment by Pablo-Aldana** [comment]
I have managed to solved the issue by adding the following dependencies into the package.json.

   "knex": "0.21.21",
    "objection": "^2.2.18",
    "pg": "8.7.1"
    
   In case this does not work I also changed temporary the NODE_ENV from production to development, and run a npm install npm start. With this the DB got created.  then put it back to production.

**Comment by PandhuWibowo** [comment]
Is it already find the solution?

**Comment by SkinnyMonk** [comment]
I have the exact same error for my ubuntu 20.04. If anyone have any proper solution, please share it here.

**Comment by smaranjit** [comment]
> Does it already find the solution?

I have upgraded the Strapi version to 4.0.5, this problem still exists.

**Comment by smaranjit** [comment]
> @ari-motors-team and @smaranjit then we need detailed reproduction steps.
> 
> A sample application or something that can be tested by us (ideally an application without much customization just native strapi stuff like models and what not, eg no uses of strapi.db in lifecycles or custom controllers).
> 
> My guess is the default configuration that some cloud hosts are applying is a conflict but without a means of reproduction there isn't much we can offer here.

hi there, please check with this sample repository to reproduce the behaviour, I hope the said issue will pop up with this sample repository, after following the steps that I have mentioned in that sample repository.

sample repository URL: https://github.com/smaranjit/strapi4-knex-issue-demonstration.git

**Comment by derrickmehaffy** [comment]
> > @ari-motors-team and @smaranjit then we need detailed reproduction steps.
> > A sample application or something that can be tested by us (ideally an application without much customization just native strapi stuff like models and what not, eg no uses of strapi.db in lifecycles or custom controllers).
> > My guess is the default configuration that some cloud hosts are applying is a conflict but without a means of reproduction there isn't much we can offer here.
> 
> hi @derrickmehaffy, please check with this sample repository to reproduce the behaviour, I hope the said issue will pop up with this sample repository, after following the steps that I have mentioned in that sample repository.
> 
> sample repository URL: https://github.com/smaranjit/strapi4-knex-issue-demonstration.git

Given I didn't use docker to setup the database and instead used my local PostgreSQL server I couldn't reproduce it so I would assume the response from the dockerized PG instance was simply too slow and Knex killed the conneciton.

**Comment by smaranjit** [comment]
> > > @ari-motors-team and @smaranjit then we need detailed reproduction steps.
> > > A sample application or something that can be tested by us (ideally an application without much customization just native strapi stuff like models and what not, eg no uses of strapi.db in lifecycles or custom controllers).
> > > My guess is the default configuration that some cloud hosts are applying is a conflict but without a means of reproduction there isn't much we can offer here.
> > 
> > 
> > hi @derrickmehaffy, please check with this sample repository to reproduce the behaviour, I hope the said issue will pop up with this sample repository, after following the steps that I have mentioned in that sample repository.
> > sample repository URL: https://github.com/smaranjit/strapi4-knex-issue-demonstration.git
> 
> Given I didn't use docker to setup the database and instead used my local PostgreSQL server I couldn't reproduce it so I would assume the response from the dockerized PG instance was simply too slow and Knex killed the conneciton.

@derrickmehaffy, I have shared the docker configuration here to quick test the problem, But I have tested the same in my locally installed Postgres instance and the same problem is happening. I have tried with Postgres max_connection configuration up to 5000, to see if the problem could be resolved, but nothing worked. Also for the docker instance - is it not supposed to work with the docker Postgres instance? I think we are missing something here, the same Starpi is working with MySQL but not with Postgres. also I can see that multiple users have almost same type of issue with Postgres.

**for reference here is my PC configuration-**
```
Intel core i7 6700k
16gb ram
256gb ssd
```

**Comment by derrickmehaffy** [comment]
> > > > @ari-motors-team and @smaranjit then we need detailed reproduction steps.
> > > > A sample application or something that can be tested by us (ideally an application without much customization just native strapi stuff like models and what not, eg no uses of strapi.db in lifecycles or custom controllers).
> > > > My guess is the default configuration that some cloud hosts are applying is a conflict but without a means of reproduction there isn't much we can offer here.
> > > 
> > > 
> > > hi @derrickmehaffy, please check with this sample repository to reproduce the behaviour, I hope the said issue will pop up with this sample repository, after following the steps that I have mentioned in that sample repository.
> > > sample repository URL: https://github.com/smaranjit/strapi4-knex-issue-demonstration.git
> > 
> > 
> > Given I didn't use docker to setup the database and instead used my local PostgreSQL server I couldn't reproduce it so I would assume the response from the dockerized PG instance was simply too slow and Knex killed the conneciton.
> 
> @derrickmehaffy, I have shared the docker configuration here to quick test the problem, But I have tested the same in my locally installed Postgres instance and the same problem is happening. I have tried with Postgres max_connection configuration up to 5000, to see if the problem could be resolved, but nothing worked. Also for the docker instance - is it not supposed to work with the docker Postgres instance? I think we are missing something here, the same Starpi is working with MySQL but not with Postgres. also I can see that multiple users have almost same type of issue with Postgres.
> 
> **for reference here is my PC configuration-**
> 
> ```
> Intel core i7 6700k
> 16gb ram
> 256gb ssd
> ```

It's not the number of connections, it's the tuning settings of how long to wait for the database to respond. Knex uses a package called tarn.js to handle the connection pooling and it has various timeout settings.

https://docs.strapi.io/developer-docs/latest/setup-deployment-guides/configurations/required/databases.html#database-pooling-options

Error message clearly says that there is a timeout getting a connection meaning the connections are timing out because the database isn't responding fast enough. I assumed others had already looked at the database tuning options since we added some of the to the documentation (we didn't write these options they are just handed straight to knex => tarn).

Best guess is if you are hitting timeouts you'll need to increase the timeout time.

**Comment by gvocale** [comment]
Running in this problem as well...

**Comment by Webbist-dev** [comment]
This is a resource issue wherever you're hosting your database. Either increase your resources for the database or increase the timeout in knex. In config/database.js add 
`pool: {
  acquireTimeoutMillis: env.int('DATABASE_POOL_ACQUIRE_TIMEOUT_MILLIS', 60000),
}`

**Comment by bublinec** [comment]
Same problem here, configuring pool does not help. Neither do added deps.

**Comment by bublinec** [comment]
In my case it was just that my **IP** was not among **trusted sources** on **DigitalOcean**.

<img width="1215" alt="image" src="https://user-images.githubusercontent.com/44405894/152928137-fd1e3585-01a5-4c6e-9986-411d6b296701.png">

**Comment by Pablo-Aldana** [comment]
> > > > > @ari-motors-team and @smaranjit then we need detailed reproduction steps.
> > > > > A sample application or something that can be tested by us (ideally an application without much customization just native strapi stuff like models and what not, eg no uses of strapi.db in lifecycles or custom controllers).
> > > > > My guess is the default configuration that some cloud hosts are applying is a conflict but without a means of reproduction there isn't much we can offer here.
> > > > 
> > > > 
> > > > hi @derrickmehaffy, please check with this sample repository to reproduce the behaviour, I hope the said issue will pop up with this sample repository, after following the steps that I have mentioned in that sample repository.
> > > > sample repository URL: https://github.com/smaranjit/strapi4-knex-issue-demonstration.git
> > > 
> > > 
> > > Given I didn't use docker to setup the database and instead used my local PostgreSQL server I couldn't reproduce it so I would assume the response from the dockerized PG instance was simply too slow and Knex killed the conneciton.
> > 
> > 
> > @derrickmehaffy, I have shared the docker configuration here to quick test the problem, But I have tested the same in my locally installed Postgres instance and the same problem is happening. I have tried with Postgres max_connection configuration up to 5000, to see if the problem could be resolved, but nothing worked. Also for the docker instance - is it not supposed to work with the docker Postgres instance? I think we are missing something here, the same Starpi is working with MySQL but not with Postgres. also I can see that multiple users have almost same type of issue with Postgres.
> > **for reference here is my PC configuration-**
> > ```
> > Intel core i7 6700k
> > 16gb ram
> > 256gb ssd
> > ```
> 
> It's not the number of connections, it's the tuning settings of how long to wait for the database to respond. Knex uses a package called tarn.js to handle the connection pooling and it has various timeout settings.
> 
> https://docs.strapi.io/developer-docs/latest/setup-deployment-guides/configurations/required/databases.html#database-pooling-options
> 
> Error message clearly says that there is a timeout getting a connection meaning the connections are timing out because the database isn't responding fast enough. I assumed others had already looked at the database tuning options since we added some of the to the documentation (we didn't write these options they are just handed straight to knex => tarn).
> 
> Best guess is if you are hitting timeouts you'll need to increase the timeout time.

It makes sense, but I wonder what has been change from 4.0.0 where this issue was not a problem and it works well with same configuration to version 4.0.1 above where this becomes an issue. 

If in 4.0.0 was working and now it isn't then I would propose to Strapi to setup the minimum connections necessary to startup as default value so that users won't have the problem by default.

**Comment by gvocale** [comment]
I keep running into this error:

```
Knex: Timeout acquiring a connection. The pool is probably full. Are you missing a .transacting(trx) call?
```
I am running strapi 4.0.7.

I have kept upgrading the performance of the strapi server, hosted on Render.com, but without luck. I am currently on the following setup:
RAM 4 GB
CPU 2

I can't reproduce the issue on localhost (not dockerized).

I have tried a bunch of method suggested on various post about this issue, but none worked successfully yet.

These are my database settings:

```tsx
// config/env/production/database.js

module.exports = ({ env }) => {
  const { host, port, database, user, password } = parse(env("DATABASE_URL"));

  return {
    connection: {
      client: "postgres",
      connection: {
        default: {
          settings: {
            charset: env("DATABASE_CHARSET", "utf8mb4"),
          },
        },
        host,
        port,
        database,
        user,
        password,
        pool: {
          min: 0,
          max: 97,
          idleTimeoutMillis: 30000000,
          createTimeoutMillis: 30000000,
          acquireTimeoutMillis: 30000000,
          propagateCreateError: false,
          ssl: {
            rejectUnauthorized: false,
          },
        },
      },
      debug: true,
    },
  };
};
```

**Comment by giveabl-st** [comment]
I'm experiencing the same issue hosting on Render.com also. Same setup as @gvocale. Works fine locally, but run into the KnexTimeoutError on Render. I can delete the Postgres DB on Render and create a new one and it will work fine, but as soon as I change a model or add a new one to Strapi and then push the code to Render, the error occurs.

**Comment by gvocale** [comment]
Thanks! I've removed `connection.default`.

I seem to have fixed the issue for now by lowering the number of max pool to 20:

Render.com has a max pool of 97, but when deploying a new release the number would be halved, since the current active deploy will be using half the pool, the new deploy the other half. I also have Pull Review url turned on, which means at the same time another deploy is active as well, which will contribute to hit the same number of pool.

For this reason I have lowered it to 20, which should allow me to have 4x deploy of strapi to run at the same time (production, new deploy, pull review ...).

So far so good, I have seen this error happen much less.

**Comment by derrickmehaffy** [comment]
It's all about tuning your config for your environment (which is why this bug report is closed, we can't do that for you :wink: )

**Comment by V-nut-V** [comment]
I tried all suggestions above, but still not working... after creating a new database and connecting with it, then switching back to the original database, it works! (why it's not working, why it works?$%^&*&^%$%^&*)

**Comment by vindhaus** [comment]
I'm not sure if this would be helpful to anyone, but I got this same error. On my non-dockerized local machine, I had a project that was using Strapi v3, and a separate project that was using Strapi v4. Two different directories. When I tried starting the project that was using v3, it was throwing me this Knex error. I had to get rid of the project using v4, and then the project using v3 started working again.

**Comment by Brocy** [comment]
Guys I'm experiencing the same issue hosting on AWS Docker + Aurora and Heroku also. Works fine locally, but run into the KnexTimeoutError. I can delete the Postgres on Heroku / AWS and create a new one and it will work fine, but as soon as I change a model or add a new one to Strapi and then push the code to Heroku / AWS, the error occurs.

Here my package.json
`{
  "name": "cv-headless-cms-v-4",
  "private": true,
  "version": "0.1.0",
  "description": "A Strapi application",
  "scripts": {
    "develop": "strapi develop",
    "start": "strapi start",
    "build": "strapi build",
    "strapi": "strapi"
  },
  "devDependencies": {},
  "dependencies": {
    "@strapi/plugin-documentation": "4.1.2",
    "@strapi/plugin-graphql": "4.1.2",
    "@strapi/plugin-i18n": "4.1.2",
    "@strapi/plugin-users-permissions": "4.1.2",
    "@strapi/provider-upload-aws-s3": "^4.1.2",
    "@strapi/provider-upload-cloudinary": "^4.1.2",
    "@strapi/strapi": "4.1.2",
    "pg": "8.6.0",
    "pg-connection-string": "^2.5.0"
  },
  "author": {
    "name": "A Strapi developer"
  },
  "strapi": {
    "uuid": "a378d6af-9d81-4821-b43a-e355ac4b33e9"
  },
  "engines": {
    "node": ">=12.x.x <=16.x.x",
    "npm": ">=6.0.0"
  },
  "license": "MIT"
} `

Notice that the pipeline run without error but no application running, in AWS I got this error.

`2022-03-07 15:23:02error Command failed with exit code 1.
2022-03-07 15:23:02info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
2022-03-07 15:23:01[2022-03-07 16:23:01.997] [31merror[39m: Knex: Timeout acquiring a connection. The pool is probably full. Are you missing a .transacting(trx) call?
2022-03-07 15:23:01KnexTimeoutError: Knex: Timeout acquiring a connection. The pool is probably full. Are you missing a .transacting(trx) call?
2022-03-07 15:23:01at Client_PG.acquireConnection (/opt/node_modules/knex/lib/client.js:295:26)
2022-03-07 15:23:01at async Runner.ensureConnection (/opt/node_modules/knex/lib/execution/runner.js:259:28)
2022-03-07 15:23:01at async Runner.run (/opt/node_modules/knex/lib/execution/runner.js:30:19)
2022-03-07 15:23:01at async PostgresqlSchemaInspector.getForeignKeys (/opt/node_modules/@strapi/database/lib/dialects/postgresql/schema-inspector.js:205:22)
2022-03-07 15:23:01at async /opt/node_modules/@strapi/database/lib/dialects/postgresql/schema-inspector.js:127:29
2022-03-07 15:23:01at async Promise.all (index 115)
2022-03-07 15:23:01at async PostgresqlSchemaInspector.getSchema (/opt/node_modules/@strapi/database/lib/dialects/postgresql/schema-inspector.js:123:21)
2022-03-07 15:23:01at async Object.syncSchema (/opt/node_modules/@strapi/database/lib/schema/index.js:51:24)
2022-03-07 15:23:01at async Strapi.bootstrap (/opt/node_modules/@strapi/strapi/lib/Strapi.js:380:5)
2022-03-07 15:23:01at async Strapi.load (/opt/node_modules/@strapi/strapi/lib/Strapi.js:408:5)
2022-03-07 15:23:01at async Strapi.start (/opt/node_modules/@strapi/strapi/lib/Strapi.js:161:9)
2022-03-07 15:23:01[2022-03-07 16:23:01.996] [34mdebug[39m: ⛔ Server wasn't able to start properly.
2022-03-07 15:21:58$ strapi start
2022-03-07 15:21:58yarn run v1.22.17`

**Comment by wmcmorrow** [comment]
Experiencing exactly the same issue. 

Worked fine locally, I made lots of changes to the models, pushed to Heroku and now I'm getting this error.

**Comment by giveabl-st** [comment]
The only thing that worked for me was increasing the hosting specs to the recommended required by Strapi (2 CPU Cores & 4 GB RAM): https://docs.strapi.io/developer-docs/latest/setup-deployment-guides/deployment.html

Anything less than those performance specs resulted in the KnexTimeoutError whenever I changed/added models. I also tried changing the connection pool settings and couldn't get it to resolve the KnexTimeoutError no matter what values I used.

It's not ideal having to spend more for hosting than a project likely requires, but after many wasted days trying to solve this, I'm just happy it's working well now.

**Comment by gvocale** [comment]
I have to say I am keep running in this issue. I'm hosting on [www.render.com](https://render.com/) as well.

My node Strapi service is running on 2 CPU and 4GB RAM.

My Postgres database is running on 2 CPU and 4GB RAM.

I'm using the following settings in `/config/env/production/database.js`

```ts
const { parse } = require("pg-connection-string");

module.exports = ({ env }) => {
  const { host, port, database, user, password } = parse(env("DATABASE_URL"));

  return {
    connection: {
      client: "postgres",
      connection: {
        host,
        port,
        database,
        user,
        password,
        ssl: {
          rejectUnauthorized: false,
        },
      },
      debug: true,
      pool: {
        min: 0,
        max: 10,
        idleTimeoutMillis: 30000000,
        createTimeoutMillis: 30000000,
        acquireTimeoutMillis: 30000000,
        propagateCreateError: false,
      },
    },
  };
};
```

**Comment by remoblaser** [comment]
Facing the same issue on GCP right now.
Doesn't really matter what kind of machine is used or that the pool settings are.

**Comment by Brocy** [comment]
The only solution for me was to use Mysql

@derrickmehaffy I suggest to open the issue, what you think?

**Comment by gvocale** [comment]
@derrickmehaffy I find having the "Discussion" tab on github very helpful, for example like next.js does https://github.com/vercel/next.js/discussions.

That way I can search via the github search within the repo for a particular issue, say `render.com database postgres knex` and find discussions about it. That way the discussion can happen, but it doesn't have to be labeled as closed like on this thread.

I know that would be a duplicate of discord or https://forum.strapi.io/, but would allow to get to answers straight here on github rather then having to google it around.

Just a suggestion 😄

**Comment by derrickmehaffy** [comment]
> @derrickmehaffy I find having the "Discussion" tab on github very helpful, for example like next.js does https://github.com/vercel/next.js/discussions.
> 
> That way I can search via the github search within the repo for a particular issue, say `render.com database postgres knex` and find discussions about it. That way the discussion can happen, but it doesn't have to be labeled as closed like on this thread.
> 
> I know that would be a duplicate of discord or https://forum.strapi.io/, but would allow to get to answers straight here on github rather then having to google it around.
> 
> Just a suggestion 😄 

We used to use discussions but they didn't have the features we wanted. Thus we created the forum: https://forum.strapi.io

**Comment by MedServ** [comment]
Adding this parameter to the configuration solved the problem for me: 
acquireConnectionTimeout: 600000,
Example:
connection: {
    client: 'postgres',
    connection: {
      host: env ('DATABASE_HOST', '127.0.0.1'),
      port: env.int ('DATABASE_PORT', 5432),
      database: env ('DATABASE_NAME', '*****'),
      user: env ('DATABASE_USERNAME', '****'),
      password: env ('DATABASE_PASSWORD', '&&&&'),
      ssl: env.bool ('DATABASE_SSL', false),
    },
    acquireConnectionTimeout: 600000,
    pool: {
      min: 2,
      max: 6,
      acquireTimeoutMillis: 300000,
      createTimeoutMillis: 300000,
      destroyTimeoutMillis: 50000,
      idleTimeoutMillis: 300000,
      reapIntervalMillis: 10000,
      createRetryIntervalMillis: 2000,
      propagateCreateError: false
    },
    debug: true
  }
});

**Comment by gvocale** [comment]
@MedServ thanks! I have just started using the same settings on render.com with a new database and so far no `Knex: Timeout` error yet... let's see...

@derrickmehaffy the `acquireConnectionTimeout` option is undocumented on the Strapi Database configuration page. If that is helping in any way mitigate this Knex: Timeout error, would it be worth to add a note to the docs?

**Comment by derrickmehaffy** [comment]
> @MedServ thanks! I have just started using the same settings on render.com with a new database and so far no `Knex: Timeout` error yet... let's see...
> 
> @derrickmehaffy the `acquireConnectionTimeout` option is undocumented on the Strapi Database configuration page. If that is helping in any way mitigate this Knex: Timeout error, would it be worth to add a note to the docs?

aquireConnectionTimeout shouldn't assist here unless the database is taking over a minute to respond to a connection attempt.

I'll remind everyone that we don't usually document every option from packages we don't maintain. All of these pool settings come from Knex.js and it's nested pool manager Tarn.js

https://github.com/Vincit/tarn.js
https://github.com/knex/knex

**Comment by gvocale** [comment]
@derrickmehaffy I seem to recall reading another thread on the Knex github page the error may be indeed a Knex shortage / bug. Somehow the connection timing out could be the very first connection made. Since Knex is the official way Strapi suggests to hook up to Postgres I would expect a bit more of support from Strapi to handle this kink of Knex. I didn't know Strapi was passing more options down to Knex other than the ones documented on the Strapi Database documentation page...  A note saying any Knex setting is going to be passed down from Strapi could be helpful. My two cents 😄

**Comment by derrickmehaffy** [comment]
> @derrickmehaffy I seem to recall reading another thread on the Knex github page the error may be indeed a Knex shortage / bug. Somehow the connection timing out could be the very first connection made. Since Knex is the official way Strapi suggests to hook up to Postgres I would expect a bit more of support from Strapi to handle this kink of Knex. I didn't know Strapi was passing more options down to Knex other than the ones documented on the Strapi Database documentation page...  A note saying any Knex setting is going to be passed down from Strapi could be helpful. My two cents 😄 

That's the thing, we aren't passing anything. And since we aren't, knex is using it's defaults. This is common is just about all software and isn't just limited to knex. We follow the same logic for koa, helmet, passport, grant/purest, all of the various SDKs for upload/email.

It's not functionally viable for us to replicate the documentation for every library we use.

**Comment by NicoHof** [comment]
@gvocale It has been a while since last activity in this issue.
I am still facing this issue and it costs a lot of time and causes frustration.
- Did you solve this problem finally?
- Do you have a workaround with your current settings? (For me sometimes just redeploying works, but sometimes even after 5 tries, not)
- Is it maybe an issue with render.com? Did you try another hosting?

@derrickmehaffy I see your point 100%. It is not the task of strapi core team to solve individual hosting issues. But render.com is a suggested hoster in the docs and its stated that 2GB ram + 2 CPU cores are enough to run strapi in production. For me it looks like that for a lot of people it is not working with this spec. (See this issue, various forum posts, ...)

For me it isnt even clear if keeping to increase the ressources on the server even solves the issue on deployment?

So IMO this issue should, at least, not be closed without further discussing/providing some troubleshooting guide on hosting services with the project.

**Comment by gvocale** [comment]
@NicoHof

I didn't solve the issue... in fact it cause lots of frustration on this side too 😢 

We finally switched from PostgreSQL to MariaDB. We don’t get anymore the `Knex: Timeout acquiring a connection`, but we get some more descriptive errors.

One bug we figure out is sometimes columns name were too long (more then 64 characters), like described into this bug #13117 . So for now we manually shorten the collectionName as described here #12101 

Another bug we’re always running into is `ER_DUP_KEYNAME: Duplicate key name` when running automatic migrations on production. So far we’ve been able to solve it by manually purging the strapi_content_types_schema key from the strapi_core_store_settings table as suggested here [Discord 1](https://discord.com/channels/811989166782021633/827318019029336166/966430371010994206)

My suspect is, when using Postgres, these two issues above return a generic `Knex: Timeout acquiring a connection` error. So for now will stick with MariaDB to hopefully get some better errors and willing to try again using PostgreSQL when someone will figure out more about the reason behind the Knex error...

**Comment by julesrenaud** [comment]
> > @MedServ thanks! I have just started using the same settings on render.com with a new database and so far no `Knex: Timeout` error yet... let's see...
> > @derrickmehaffy the `acquireConnectionTimeout` option is undocumented on the Strapi Database configuration page. If that is helping in any way mitigate this Knex: Timeout error, would it be worth to add a note to the docs?
> 
> aquireConnectionTimeout shouldn't assist here unless the database is taking over a minute to respond to a connection attempt.
> 
> I'll remind everyone that we don't usually document every option from packages we don't maintain. All of these pool settings come from Knex.js and it's nested pool manager Tarn.js
> 
> https://github.com/Vincit/tarn.js https://github.com/knex/knex

That's how I fixed the same issue. This was a total production showstopper and a very obscure solution, so it would be wise to add a note about it in the official Strapi doc. Resources per say aren't at play here, because I went up to 4 CPU / 8GB RAM on both Strapi's and postgres' servers. The issues occurs on Strapi installations that have a lot of tables and relations. When Strapi goes in to update schemas in postgres it very quickly uses up all available connections in the pool and the app obviously crashes.

It seems pretty obvious there is an issue on Strapi's side with regards to how it goes about updating tables in postgres. The issue happens only when there are a lot of tables. I don't see this as a tuning issue since the solution is to give knex more than a minute to acquire a connection, which is an unreasonable amount of time for such an operation.

**Comment by derrickmehaffy** [comment]
@NicoHof 

> So IMO this issue should, at least, not be closed without further discussing/providing some troubleshooting guide on hosting services with the project.

The problem is that we don't consider the issue to be a bug, so I agree further discussion is required but for that type of discussion it should happen in the community spaces like on the forum: https://forum.strapi.io/t/strapi-postgres-connection-pool-breaks-connection/17926/5

GitHub issues should be for bugs where a PR could/should fix the problem. I don't believe that is the case here.

**Comment by nicoworq** [comment]
Was having the same issue, but on MySQL. After dealing with it for about 1 hour, just a simple MySQL server restart do the job 🤣🤣

**Comment by ddrosatosimode** [comment]
I was having the same error after i added new field to existing content type. Removing the new field opens the admin again.
Its a huge issue generally and i dont think its knex only related. Strapi version 4.1.8 on Heroku with Postgr. Pool options in database.js dont resolve the issue.

Honestly i dont see the reason behind these issues. I cant understand why a slight change of a content type schemas can start a knex error.

**Comment by NicoHof** [comment]
> That's how I fixed the same issue. This was a total production showstopper and a very obscure solution, so it would be wise to add a note about it in the official Strapi doc. Resources per say aren't at play here, because I went up to 4 CPU / 8GB RAM on both Strapi's and postgres' servers. The issues occurs on Strapi installations that have a lot of tables and relations. When Strapi goes in to update schemas in postgres it very quickly uses up all available connections in the pool and the app obviously crashes.
> 
> It seems pretty obvious there is an issue on Strapi's side with regards to how it goes about updating tables in postgres. The issue happens only when there are a lot of tables. I don't see this as a tuning issue since the solution is to give knex more than a minute to acquire a connection, which is an unreasonable amount of time for such an operation.

I had a lot of knex issues on render.com, so I switched to digitalocean app platform and (for me) all the issues are gone.
I dont know if they come back once my database has more entries but for now everything works good. So maybe worth a try for other who experience issues on render.com

See this discussion https://forum.strapi.io/t/strapi-postgres-connection-pool-breaks-connection/17926/20

**Comment by MichaelFrieze** [comment]
I am also not having any issues with I deploy to Digital Ocean. 

But, when I deploy to Railway, Fly, and Render, I get this issue. I really want this to work on Fly, so maybe I will try using planetscale's MySQL service.

**Comment by Brianggalvez** [comment]
I'm having this issue on DO too. It looks like the query that causes the timeout is the following one: 
![image](https://user-images.githubusercontent.com/18324872/172012476-cd5a4080-e691-4983-a0ff-8b959f3ec855.png)

**Comment by rafaelbm9126** [comment]
I don't know if it works for you

![knex_strapi](https://user-images.githubusercontent.com/22375348/172921871-f4a6699e-4488-4d31-ac2b-61b059d9a1ee.png)

**Comment by derrickmehaffy** [comment]
Something we noticed with another customer yesterday was an undocumented timeout setting from Knex.js. I've submitted the PR to the documentation that might help some of you.

PR: https://github.com/strapi/documentation/pull/929
Temp Doc link: https://documentation-git-fix-missingtimeoutdb-strapijs.vercel.app/developer-docs/latest/setup-deployment-guides/configurations/required/databases.html#connection-configuration-object

This doesn't "fix" the problem but acts more like a band-aid as the default will drop any connection that takes longer than 60 seconds meaning there is still an issue with some of your environments on why something would take longer than 60 seconds but increasing this timeout can help give it some more time to finish whatever it was doing.

Basically you apply it like this:

```js
module.exports = ({ env }) => ({
  connection: {
    client: 'whatever you are using',
    connection: {
      // ...
    },
    debug: false,
    acquireConnectionTimeout: 300000, // 5 minute timeout
  },
});
```

**Comment by derrickmehaffy** [comment]
> I work at Render and we've been seeing upticks in this error, we've also seen reports of the same thing occurring on Fly and Railway.
> 
> The only way we've seen to fix it is to increase the `acquireTimeoutMillis` setting as highlighted above which in our experience has been 100% successful at eliminating this issue.

@johnbeynon I've been working with our DevRel team to look at some Render issues also and I've noticed some odd problems with the managed databases offered. I'm wondering what the specs are for those and if they are running over a high latency connection.

**Comment by johnbeynon** [comment]
@derrickmehaffy no, they are in the same K8s namespace as the services accessing them (assuming they are on Render) and all within the same AWS region.

**Comment by derrickmehaffy** [comment]
> @derrickmehaffy no, they are in the same K8s namespace as the services accessing them (assuming they are on Render) and all within the same AWS region.

I'm wondering if it's the resource allocation :thinking:, typically I would advise to anyone to **never** dockerize/containerize a database for any meaningful purpose and in the same respect I've never been able to reproduce this issue using RDS which is likely why because I never tested on a k8s deployed database in a production-like env.

**Comment by johnbeynon** [comment]
The thing is, the error that we saw was often after sites had been left not being accessed for periods of time and it seemed like the database connection had been dropped and then the next request fails because the DB connection had dropped. 

It would manifest as:

```

**Comment by johnbeynon** [comment]
The thing is, the error that customers saw was often after sites had been left not being accessed for periods of time and it seemed like the database connection had been dropped and then the next request fails because the DB connection had dropped. 

It would manifest as:

```
May 31 11:12:01 AM [2022-05-31 10:12:01.823] debug: ⛔️ Server wasn't able to start properly.

May 31 11:12:01 AM [2022-05-31 10:12:01.824] error: select "t0"."id" from "admin_users" as "t0" limit $1 - Connection terminated unexpectedly

May 31 11:12:01 AM error Command failed with exit code 1.

May 31 11:12:01 AM Error: select "t0"."id" from "admin_users" as "t0" limit $1 - Connection terminated unexpectedly at Connection.<anonymous>
```

but that first line, `Server wasn't able to start properly.` was often many hours after the server had started, processed requests and then sat idle for a period of time.

There was never any 60s to obtain a connection or anything. 

I appreciate your point of DBs in K8s, but that's the foundation of our business and the likes of Railway and Fly - we've not experienced this across other frameworks, only Strapi.

**Comment by derrickmehaffy** [comment]
In a case like this I believe the last time it was looked into, docker users need to set the pool min option to 0 instead of 1. Given our distribution of users docker users are in the minority so our default here is one but docker does some weird stuff in their network code that will kill idle connections after a period of time.

---

No trying to drag on Render with regards to dockerized DB's more so a matter of bias opinion there. I've seen some extremely bad cases of poor implementation of this which revolved around massive performance problems at scale and higher than average issues with data corruption and loss. Either way not a topic for this issue.

**Comment by julesrenaud** [comment]
> I work at Render and we've been seeing upticks in this error, we've also seen reports of the same thing occurring on Fly and Railway.
> 
> The only way we've seen to fix it is to increase the `acquireTimeoutMillis` setting as highlighted above which in our experience has been 100% successful at eliminating this issue.

I'm having the same issue with AWS RDS (connecting from a docker container running on a EC2 instance)

**Comment by bencehusi** [comment]
I have the same issue. I've tried to increase acquireTimeoutMillis but my connection attempt still breaks after 60000ms.
Any idea?

This is my config (Strapi v 4.2.3)
```
module.exports = ({ env }) => ({
  connection: {
    client: 'postgres',
    connection: {
      host: env('DATABASE_HOST'),
      port: env.int('DATABASE_PORT'),
      database: env('DATABASE_NAME'),
      user: env('DATABASE_USERNAME'),
      password: env('DATABASE_PASSWORD'),
      ssl: env.bool('DATABASE_SSL'),
    },
    pool: {
      min: 0,  // default 2
      max: 6,  // default 10
      acquireTimeoutMillis: 120000, // THIS DOESN'T INCREASE THE TIMEOUT!!!! WHAT AM I DOING WRONG?
    },
  },
});

```

**Comment by julesrenaud** [comment]
@bencehusi the key for me was using `acquireConnectionTimeout: 600000` outside the `pool`config. Fixed my issue completely, nothing else did. Here's a sample config:

```
connection: {
    client: 'postgres',
    connection: {
      host: env('DATABASE_HOST'),
      port: env.int('DATABASE_PORT'),
      database: env('DATABASE_NAME'),
      user: env('DATABASE_USERNAME'),
      password: env('DATABASE_PASSWORD'),
      ssl: {
        rejectUnauthorized: env.bool('DATABASE_SSL_SELF', false),
      },
    },
    debug: false,
    acquireConnectionTimeout: 600000,
    pool: {
      min: 0,
      max: 100,
      acquireTimeoutMillis: 300000,
      createTimeoutMillis: 300000,
      destroyTimeoutMillis: 50000,
      idleTimeoutMillis: 300000,
      reapIntervalMillis: 10000,
      createRetryIntervalMillis: 2000,
      propagateCreateError: false,
    },
  },
```

**Comment by bencehusi** [comment]
Thanks @julesrenaud .

I've also found something that helps, so I'm posting it here:

I am using Supabase for my database and storage needs and I've contacted their customer service about this issue. They suggested that I use another port where they're providing Pgbouncer.

**So yeah: try Pgbouncer!**

And that not just resolved this issue but also helped to reduce database connection time (thus server restarts).

So, if anyone is looking for a low-hanging fruit solution I recommend Pgbouncer.

**Comment by wagneraraujo** [comment]
> @bencehusia chave para mim estava usando `acquireConnectionTimeout: 600000`fora da `pool`configuração. Corrigido meu problema completamente, nada mais fez. Aqui está uma configuração de exemplo:
> 
> ```
> connection: {
>     client: 'postgres',
>     connection: {
>       host: env('DATABASE_HOST'),
>       port: env.int('DATABASE_PORT'),
>       database: env('DATABASE_NAME'),
>       user: env('DATABASE_USERNAME'),
>       password: env('DATABASE_PASSWORD'),
>       ssl: {
>         rejectUnauthorized: env.bool('DATABASE_SSL_SELF', false),
>       },
>     },
>     debug: false,
>     acquireConnectionTimeout: 600000,
>     pool: {
>       min: 0,
>       max: 100,
>       acquireTimeoutMillis: 300000,
>       createTimeoutMillis: 300000,
>       destroyTimeoutMillis: 50000,
>       idleTimeoutMillis: 300000,
>       reapIntervalMillis: 10000,
>       createRetryIntervalMillis: 2000,
>       propagateCreateError: false,
>     },
>   },
> ```
Thanks brother, working for me!

**Comment by prodkt** [comment]
> @bencehusi the key for me was using `acquireConnectionTimeout: 600000` outside the `pool`config. Fixed my issue completely, nothing else did. Here's a sample config:
> 
> ```
> connection: {
>     client: 'postgres',
>     connection: {
>       host: env('DATABASE_HOST'),
>       port: env.int('DATABASE_PORT'),
>       database: env('DATABASE_NAME'),
>       user: env('DATABASE_USERNAME'),
>       password: env('DATABASE_PASSWORD'),
>       ssl: {
>         rejectUnauthorized: env.bool('DATABASE_SSL_SELF', false),
>       },
>     },
>     debug: false,
>     acquireConnectionTimeout: 600000,
>     pool: {
>       min: 0,
>       max: 100,
>       acquireTimeoutMillis: 300000,
>       createTimeoutMillis: 300000,
>       destroyTimeoutMillis: 50000,
>       idleTimeoutMillis: 300000,
>       reapIntervalMillis: 10000,
>       createRetryIntervalMillis: 2000,
>       propagateCreateError: false,
>     },
>   },
> ```

This was the key for me. I got thrown one error missing APP_KEYS variables. I just hadn't generated them yet. Did that in 15 seconds tossed into in my .env and I was back online. Thanks man!

**Comment by maherbeg** [comment]
For user's running on Render, we've got a fix that can fix many of the connection issues that we need to apply to your database. You'll also need to use the client side configuration settings in https://github.com/strapi/strapi/issues/11860#issuecomment-1236132786 as well to resolve additional failure modes on the client side.

**Comment by AndreaCoppi** [comment]
I have the same issue with Mysql, anyone know how to solve it?

**Comment by valentyn-dasburo** [comment]
This error might indicate problems connecting to DB. For example, if you have it when run `strapi config:dump`, you might need to provide a proper ENV_PATH variable to get the right DB connection: `ENV_PATH=./.env.dev strapi config:dump`

**Comment by PezCoder** [comment]
For those who are stuck after reading this, the following solved it for me: 🎉 
https://stackoverflow.com/questions/52324170/aws-rds-for-postgresql-cannot-be-connected-after-several-hours/52346331#52346331

It was more of an AWS issue rather than a database config issue. Found this solution here:
https://github.com/strapi/strapi/issues/10886#issuecomment-1324343901

**Comment by moadw** [comment]
This was an extremely annoying issue. After 24 hours working on it I found that the problem is not Strapi either Postgres. It's the security group in amazon that wasn't allowing DB connection. The timeout is not a pool issue but a connection block issue although node reports knex and other problems.

**Comment by hungnv1312** [comment]
lmao this bug still exists xD for strapi + render

**Comment by hungnv1312** [comment]
after deploy on render a month.... this bug back to log xD
![Uploading Screenshot 2024-02-02 002957.png…]()

**Comment by hungnv1312** [comment]
after deploy on render a month.... this bug back to log xD
![Uploading Screenshot 2024-02-02 002957.png…]()

**Comment by hungnv1312** [comment]
after deploy on render a month.... this bug back to log xD


```


==> No open HTTP ports detected
==> https://render.com/docs/web-services#host-and-port-configuration
[2024-02-01 17:25:50.781] debug: ⛔️ Server wasn't able to start properly.
[2024-02-01 17:25:50.782] error: Knex: Timeout acquiring a connection. The pool is probably full. Are you missing a .transacting(trx) call?
KnexTimeoutError: Knex: Timeout acquiring a connection. The pool is probably full. Are you missing a .transacting(trx) call?
    at Client_PG.acquireConnection (/opt/render/project/src/node_modules/@strapi/database/node_modules/knex/lib/client.js:332:26)
    at async Runner.ensureConnection (/opt/render/project/src/node_modules/@strapi/database/node_modules/knex/lib/execution/runner.js:305:28)
    at async Runner.run (/opt/render/project/src/node_modules/@strapi/database/node_modules/knex/lib/execution/runner.js:30:19)
    at async Strapi.bootstrap (/opt/render/project/src/node_modules/@strapi/strapi/dist/Strapi.js:390:9)
    at async Strapi.load (/opt/render/project/src/node_modules/@strapi/strapi/dist/Strapi.js:425:5)
    at async Strapi.start (/opt/render/project/src/node_modules/@strapi/strapi/dist/Strapi.js:216:9)

```

**Comment by Virtual-Bikram** [comment]
did someone found solution to this issue ? I am able to connect to MySQL but not with postgres ?
What configuration changes I need to make @strapi-bot @strapi-cla

**Comment by hungnv1312** [comment]
Seems to be simpler to deploy to a private VPS? Because it's the same as running on the local machine but online global on VPS
?

**Comment by xmak** [comment]
> acquireTimeoutMillis: 300000,
>       createTimeoutMillis: 300000,
>       destroyTimeoutMillis: 50000,
>       idleTimeoutMillis: 300000,
>       reapIntervalMillis: 10000,
>       createRetryIntervalMillis: 2000,
>       propagateCreateError: false,

Did you setup this in your local Strapi instance or in production? Which environment is relevant for the transfer?

**Comment by xmak** [comment]
When using the "--from" option does strapi transfer require access to the database of the remote Strapi installation?

I'm running strapi transfer on my local machine with the --from option to transfer the database from the production to my local machine. The production is hosted on Azure where we use Mysql but since the environment variables on my machine are set up for accessing my development database, the transfer has no way of knowing how to connect to the remote database. Is this relevant?

**Comment by Roshan-Vampire1993** [comment]
> @bencehusi the key for me was using `acquireConnectionTimeout: 600000` outside the `pool`config. Fixed my issue completely, nothing else did. Here's a sample config:
> 
> ```
> connection: {
>     client: 'postgres',
>     connection: {
>       host: env('DATABASE_HOST'),
>       port: env.int('DATABASE_PORT'),
>       database: env('DATABASE_NAME'),
>       user: env('DATABASE_USERNAME'),
>       password: env('DATABASE_PASSWORD'),
>       ssl: {
>         rejectUnauthorized: env.bool('DATABASE_SSL_SELF', false),
>       },
>     },
>     debug: false,
>     acquireConnectionTimeout: 600000,
>     pool: {
>       min: 0,
>       max: 100,
>       acquireTimeoutMillis: 300000,
>       createTimeoutMillis: 300000,
>       destroyTimeoutMillis: 50000,
>       idleTimeoutMillis: 300000,
>       reapIntervalMillis: 10000,
>       createRetryIntervalMillis: 2000,
>       propagateCreateError: false,
>     },
>   },
> ```

I tried like this, but still facing the problem every time I make changes to any content-type schema.
The error is being thrown when strapi tries to create audit logs. Are there any updates about this problem?

**Comment by hassan-qureshi1** [comment]
anyone has fixed this issue?

**Comment by Kabilan2370** [comment]
https://github.com/Kabilan2370/Error-Knex-Timeout-acquiring-a-connection.-The-pool-is-probably-full.git

I faced this issue for over a week and tried multiple approaches to resolve it. Eventually, a small change fixed the problem, Take a minute look into this repo
